redis.go

```diff
@@ -1,46 +1,199 @@
 }
 
-// Set the given key with the given value and expiration time.
-func (c *Client) Set(key, value string, exp time.Duration) {
-	c.client.Set(c.client.Context(), key, value, exp)
+// Exports implements the modules.Instance interface and returns
+// the exports of the JS module.
+func (mi *ModuleInstance) Exports() modules.Exports {
+	return modules.Exports{Named: map[string]interface{}{
+		"Client": mi.NewClient,
+	}}
 }
 
-// Get returns the value for the given key.
-func (c *Client) Get(key string) (string, error) {
-	res, err := c.client.Get(c.client.Context(), key).Result()
+// NewClient is the JS constructor for the redis Client.
+//
+// Under the hood, the redis.UniversalClient will be used. The universal client
+// supports failover/sentinel, cluster and single-node modes. Depending on the options,
+// the internal universal client instance will be one of those.
+//
+// The type of the underlying client depends on the following conditions:
+// 1. If the MasterName option is specified, a sentinel-backed FailoverClient is used.
+// 2. if the number of Addrs is two or more, a ClusterClient is used.
+// 3. Otherwise, a single-node Client is used.
+func (mi *ModuleInstance) NewClient(call goja.ConstructorCall) *goja.Object {
+	rt := mi.vu.Runtime()
+
+	var optionsArg map[string]interface{}
+	err := rt.ExportTo(call.Arguments[0], &optionsArg)
+	if err != nil {
+		common.Throw(rt, errors.New("unable to parse options object"))
+	}
+
+	opts, err := newOptionsFrom(optionsArg)
 	if err != nil {
-		return "", err
+		common.Throw(rt, fmt.Errorf("invalid options; reason: %w", err))
+	}
+
+	redisOptions := &redis.UniversalOptions{
+		Addrs:              opts.Addrs,
+		DB:                 opts.DB,
+		Username:           opts.Username,
+		Password:           opts.Password,
+		SentinelUsername:   opts.SentinelUsername,
+		SentinelPassword:   opts.SentinelPassword,
+		MasterName:         opts.MasterName,
+		MaxRetries:         opts.MaxRetries,
+		MinRetryBackoff:    opts.MinRetryBackoff,
+		MaxRetryBackoff:    opts.MaxRetryBackoff,
+		DialTimeout:        opts.DialTimeout,
+		ReadTimeout:        opts.ReadTimeout,
+		WriteTimeout:       opts.WriteTimeout,
+		PoolSize:           opts.PoolSize,
+		MinIdleConns:       opts.MinIdleConns,
+		MaxConnAge:         opts.MaxConnAge,
+		PoolTimeout:        opts.PoolTimeout,
+		IdleTimeout:        opts.IdleTimeout,
+		IdleCheckFrequency: opts.IdleCheckFrequency,
+		MaxRedirects:       opts.MaxRedirects,
+		ReadOnly:           opts.ReadOnly,
+		RouteByLatency:     opts.RouteByLatency,
+		RouteRandomly:      opts.RouteRandomly,
+	}
+
+	vuState := mi.vu.State()
+	if vuState != nil {
+		// If k6 has a TLSConfig set in its state, use
+		// it has redis' client TLSConfig too.
+		if vuState.TLSConfig != nil {
+			redisOptions.TLSConfig = vuState.TLSConfig
+		}
+
+		// use k6's lib.DialerContexter function has redis'
+		// client Dialer
+		redisOptions.Dialer = vuState.Dialer.DialContext
+	}
+
```

Comment:
I would argue that given this we should only allow this to happen in the vu context.

Especially as k6 has a general rule of no network requests in the init context and I guess `redis.NewUniversalClient` is going to at least connect to redis

Reply:
What would be the canonical way to ensure that's the case?

I'm gonna look into the redis' client code to verify if it actually opens any connections when initializing it :+1:

Reply:
Just enforcing that anything that can do it gets an error if it's not in the vu context as in [here](https://github.com/grafana/k6/blob/556747fbe585c981507904d634e705c68891e663/js/modules/k6/metrics/metrics.go#L96-L100) for example. In this example this protects so that metric samples being emitted only in vu context but, it's the same idea.

Reply:
I've made sure the constructor `new redis.Client` does not produce any IO, but simply sets the options. Moved the connection step in a dedicated `Connect` method, which uses the options to initialize the client's connection using the VU dialer and `TLSConfig`. Also ensured that connecting in the init context throws a `common.InitContextError`. Thank you for the pointer 🙇🏻

Reply:
I ended up hiding the call to `.connect` in the different commands directly. Thus, makinga redis client operation that opens a connection in the init context will still throw, but the connection will be setup automatically under the hood whenever using one of the redis client's methods.